libraries/chain/db_update.cpp

```diff
@@ -217,38 +231,35 @@ bool database::check_for_blackswan( const asset_object& mia, bool enable_black_s
     // stop when limit orders are selling too little USD for too much CORE
     auto lowest_possible_bid  = price::min( mia.id, bitasset.options.short_backing_asset );
 
-    assert( highest_possible_bid.base.asset_id == lowest_possible_bid.base.asset_id );
+    FC_ASSERT( highest_possible_bid.base.asset_id == lowest_possible_bid.base.asset_id );
     // NOTE limit_price_index is sorted from greatest to least
     auto limit_itr = limit_price_index.lower_bound( highest_possible_bid );
     auto limit_end = limit_price_index.upper_bound( lowest_possible_bid );
 
-    auto call_min = price::min( bitasset.options.short_backing_asset, mia.id );
-    auto call_max = price::max( bitasset.options.short_backing_asset, mia.id );
-    auto call_itr = call_price_index.lower_bound( call_min );
-    auto call_end = call_price_index.upper_bound( call_max );
-
-    if( call_itr == call_end ) return false;  // no call orders
-
-    price highest = settle_price;
     if( limit_itr != limit_end ) {
-       assert( settle_price.base.asset_id == limit_itr->sell_price.base.asset_id );
-       highest = std::max( limit_itr->sell_price, settle_price );
+       FC_ASSERT( highest.base.asset_id == limit_itr->sell_price.base.asset_id );
+       highest = std::max( limit_itr->sell_price, highest );
     }
 
     auto least_collateral = call_itr->collateralization();
     if( ~least_collateral >= highest  ) 
     {
+       wdump( (*call_itr) );
        elog( "Black Swan detected: \n"
              "   Least collateralized call: ${lc}  ${~lc}\n"
            //  "   Highest Bid:               ${hb}  ${~hb}\n"
-             "   Settle Price:              ${sp}  ${~sp}\n"
-             "   Max:                       ${h}   ${~h}\n",
+             "   Settle Price:              ${~sp}  ${sp}\n"
+             "   Max:                       ${~h}   ${h}\n",
             ("lc",least_collateral.to_real())("~lc",(~least_collateral).to_real())
           //  ("hb",limit_itr->sell_price.to_real())("~hb",(~limit_itr->sell_price).to_real())
             ("sp",settle_price.to_real())("~sp",(~settle_price).to_real())
             ("h",highest.to_real())("~h",(~highest).to_real()) );
        FC_ASSERT( enable_black_swan, "Black swan was detected during a margin update which is not allowed to trigger a blackswan" );
-       globally_settle_asset(mia, ~least_collateral );
+       if( maint_time > HARDFORK_CORE_338_TIME && ~least_collateral <= settle_price )
+          // global settle at feed price if possible
+          globally_settle_asset(mia, settle_price );
+       else
+          globally_settle_asset(mia, ~least_collateral );
```

Comment:
Fixing #338 means when matching a call order with a limit order, if the call order is maker, match price should always be MSSP. If matching at MSSP would trigger a black swan event, rather than trigger it later (on match), we trigger it here (when there is a price feed change). There is a comment added above:
```
+    if( maint_time > HARDFORK_CORE_338_TIME )
+       // due to #338, we won't check for black swan on incoming limit order, so need to check with MSSP here
+       highest = bitasset.current_feed.max_short_squeeze_price();
```

In this case, if LCR (CR of call order with least CR) is above 100%, we settle at 100%, otherwise settle at LCR.

Reply:
By definition, after a black swan the asset is always settled at the CR of the least collateralized call. This should not be changed, or be approved by shareholders explicitly.

Reply:
Here it settles at feed price:
https://github.com/bitshares/bitshares-core/blob/2.0.180202/libraries/chain/db_market.cpp#L602-L609
```
       if( usd_to_buy * match_price > call_itr->get_collateral() )
       {
          elog( "black swan detected" ); 
          edump((enable_black_swan));
          FC_ASSERT( enable_black_swan );
          globally_settle_asset(mia, bitasset.current_feed.settlement_price );
          return true;
       }
```

IMHO it makes sense to settle at feed price if the positions have more collateral than the debt worth, settling at least CR is only because unable to settle at feed price.

Reply:
Interesting. Wrt https://bitshares.org/technology/price-stable-cryptocurrencies/ I would consider this a bug.
> If this revaluation happens faster than the short positions can be forced to cover, then all SmartCoins are liquidated at the exchange rate of the least collateralized short position.

Whether it makes sense or not, it *is* a change of behaviour.

Reply:
Also there is this part in the document linked above:
>In the event that the least-collateralized short position lacks enough collateral to cover at the price feed, then all BitUSD positions are automatically force settled at the price of the least collateralized short.

IMHO the document needs to be revised, it doesn't mention margin calls or MSSR etc at all, at least it's incomplete. Due to margin calls, global force settlement can happen when the short positions still have more than 100% collateral.

I'll draft a new BSIP and ask stake holders to vote.